spectator-ext-aws/src/main/java/com/netflix/spectator/aws/SpectatorRequestMetricCollector.java

```diff
@@ -72,37 +72,48 @@
   };
 
   private final Registry registry;
+  private final Map<String, String> customTags;
 
   /**
    * Constructs a new instance.
    */
   public SpectatorRequestMetricCollector(Registry registry) {
+    this(registry, Collections.emptyMap());
+  }
+
+  /**
+   * Constructs a new instance. Custom tags provided by the user will be applied to every metric.
+   * If there is a conflict between a tag added by SpectatorRequestMetricCollector and a custom tag,
+   * the custom tag will override the built-in one.
```

Comment:
Why is this desirable? Typically the built-in values are extracted from the context of the request, since you cannot know this for a static map, it seems strange to allow that sort of override. Perhaps a better option would be a static check that would reject custom tags that try to use a key that is provided by default.

Reply:
I don't see a current use case for it, but the idea was to provide some sort of extensibility, where a user could choose to override an existing tag to eg repurpose it for something else.
If that's not something we want, the static check makes sense.

Reply:
You know, if we added a static check, it makes it more difficult to add tags in the future, because we could break people at runtime. That doesn't seem like a good idea. I think we should pick a default - have the user's tags override ours or vice versa. My thoughts on this are that if a user is trying to override one of our tags then we should let them, so I think the former makes sense. What do you think?

Reply:
Permitting customization at all means we potentially break users if the default set of tags from the extension changes. If allowing customization, it is more of a question of how things break. In many cases where there is a concern for the fall out of runtime errors the behavior is to log by default, but throw for test cases:

http://netflix.github.io/spectator/en/latest/intro/testing/#exceptions

The goal is to make it more likely the developer will notice early on and at least log so that it is not completely silent when something is misbehaving.

Regarding overrides, in my opinion a user shouldn't override any of the provided tags because it causes confusion and makes it harder to reason about aggregate behavior using these metrics. For example, if the user overrides the `requestType`, then everyone querying it would need to know that or they could be confused that they are not seeing the contribution for a subset of the requests because the developer chose to put some arbitrary value there.

Reply:
Synced up in person, makes sense, agreed. Thanks for pointing me at Registry#propagate 👍